### PR TITLE
custom hashCode/== in Element to speed up inheritFromWidgetOfExactType

### PR DESCRIPTION
**Pull request by yjbanov**
See https://github.com/dart-lang/sdk/issues/31011
**WARNING**: There's an ongoing regression introduced by https://github.com/flutter/flutter/pull/12933

### CODE REVIEW

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -2507,6 +2507,21 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
 
   Element _parent;
 
+  // Custom implementation of `operator ==` optimized for the ".of" pattern
+  // used with `InheritedWidgets`.
+  @override
+  bool operator ==(Object other) => identical(this, other);
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: is that not what the inherited operator == does?

File: packages/flutter/lib/src/widgets/framework.dart [hunk 2]
```diff
@@ -2507,6 +2507,21 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
...
+
+  // Custom implementation of hash code optimized for the ".of" pattern
+  // used with `InheritedWidgets`.
+  //
+  // `Element.inheritFromWidgetOfExactType` relies heavily on hash-based
+  // `Set` look-ups, putting this getter on the performance critical path.
+  @override
+  int get hashCode => _cachedHash;
+  final int _cachedHash = _nextHashCode = (_nextHashCode + 1) % 0xffffff;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: why only 24 bits?

**Comment by Hixie** [comment]
LGTM

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (yjbanov) [reply packages/flutter/lib/src/widgets/framework.dart]: I believe so, but the linter says if you overload `hashCode` you must also overload `operator==`.

File: packages/flutter/lib/src/widgets/framework.dart (hunk 2 above)
Comment (yjbanov) [reply packages/flutter/lib/src/widgets/framework.dart]: Copied from https://github.com/dart-lang/sdk/issues/31011#issuecomment-344245320, but I'm guessing this is so it fits within the [SMI representation](https://www.dartlang.org/articles/dart-vm/numeric-computation).

Comment (yjbanov) [reply packages/flutter/lib/src/widgets/framework.dart]: Adding a comment with explanation.

Comment (Hixie) [reply packages/flutter/lib/src/widgets/framework.dart]: We're about to switch to 64bit ints, which might make this moot.
cc @floitschG for advice.

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/framework.dart]: fair enough. i guess this is no worse than an `// ignore` would be.

**Pull request merged**